IdentityCore/src/webview/embeddedWebview/MSIDOAuth2WebviewController.m

```diff
@@ -26,12 +26,185 @@
+    }
+    
+    [_webviewUIController stop:^{[self dispatchCompletionBlock:error URL:endURL];}];
+    _webviewUIController = nil;
+    
+    return YES;
+}
+
+- (void)dispatchCompletionBlock:(NSError *)error URL:(NSURL *)url
+{
+    // NOTE: It is possible that competition between a successful completion
+    //       and the user cancelling the authentication dialog can
+    //       occur causing this method to be called twice. The competition
+    //       cannot be blocked at its root, and so this method must
+    //       be resilient to this condition and should not generate
+    //       two callbacks.
+    [_completionLock lock];
+    
+    if ( _completionHandler )
+    {
+        void (^completionHandler)(MSIDWebOAuth2Response *response, NSError *error) = _completionHandler;
+        _completionHandler = nil;
+        
+        //TODO: think about how to generate different type of response from URL
+        MSIDWebOAuth2Response *response = [MSIDWebOAuth2Response new];
+        dispatch_async( dispatch_get_main_queue(), ^{
+            completionHandler(response, error);
+        });
+    }
+    
+    [_completionLock unlock];
+}
+
+#pragma mark - MSIDWebviewDelegate
+
+- (void)webAuthDidStartLoad:(NSURL*)url
+{
+}
+
+- (void)webAuthDidFinishLoad:(NSURL*)url
+{
+}
+
+- (BOOL)webAuthShouldStartLoadRequest:(NSURLRequest *)request
+{
+    NSString *requestURL = [request.URL absoluteString];
+    
+    // Stop at the end URL.
+    if ([[requestURL lowercaseString] hasPrefix:[_webviewRequest.endURL.absoluteString lowercaseString]])
+    {
+        NSURL* url = request.URL;
+        [self webAuthDidCompleteWithURL:url];
+        
+        // Tell the web view that this URL should not be loaded.
+        return NO;
+    }
+    
+    return YES;
+}
+
+// The user cancelled authentication
+- (void)webAuthDidCancel
+{
+    // Dispatch the completion block
+    NSError *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorUserCancel, @"The user has cancelled the authorization.", nil, nil, nil, nil, nil);
+    [self endWebAuthenticationWithError:error orURL:nil];
+}
+
+// Authentication completed at the end URL
+- (void)webAuthDidCompleteWithURL:(NSURL *)endURL
+{
+    [self endWebAuthenticationWithError:nil orURL:endURL];
+}
+
+// Authentication failed somewhere
+- (void)webAuthDidFailWithError:(NSError *)error
+{
+    // Ignore WebKitError 102 for OAuth 2.0 flow.
+    if ([error.domain isEqualToString:@"WebKitErrorDomain"] && error.code == 102)
+    {
+        return;
+    }
+    
+    // Prior to iOS 10 the WebView trapped out this error code and didn't pass it along to us
+    // now we have to trap it out ourselves.
+    if ([error.domain isEqualToString:NSCocoaErrorDomain] && error.code == NSUserCancelledError)
+    {
+        return;
+    }
+    
+    // If we failed on an invalid URL check to see if it matches our end URL
+    if ([error.domain isEqualToString:@"NSURLErrorDomain"] && (error.code == -1002 || error.code == -1003))
```

Comment:
This part is removed as previously in ADAL, we catch this error to check:
1) whether url matches `endURL`;
2) whether it is a PKeyAuth challenge in URL form.

Since 
1) we already do `endURL` matching before a url is loading;
2) we don't do PKeyAuth anymore,

So I removed the whole if block.